pkg/builder/naive_build_directory.go

```diff
@@ -56,36 +60,67 @@ func (d *naiveBuildDirectory) mergeDirectoryContents(ctx context.Context, digest
+		childComponents = append(childComponents, file.Name)
 		childDigest, err := digest.NewDerivedDigest(file.Digest)
 		if err != nil {
 			return util.StatusWrapf(err, "Failed to extract digest for input file %#v", path.Join(childComponents...))
 		}
-		if err := d.contentAddressableStorage.GetFile(ctx, childDigest, inputDirectory, file.Name, file.IsExecutable); err != nil {
-			return util.StatusWrapf(err, "Failed to obtain input file %#v", path.Join(childComponents...))
-		}
+		// Extract values out of object so we don't end up with stale objects when we pass into go routine.
+		fileName := file.Name
+		fileIsExecutable := file.IsExecutable
+
+		actionGroup = append(actionGroup, func(actionContext *re_util.ActionContext) {
+			if err := d.contentAddressableStorage.GetFile(actionContext.Ctx(), childDigest, inputDirectory, fileName, fileIsExecutable); err != nil {
+				actionContext.Done(util.StatusWrapf(err, "Failed to obtain input file %#v", path.Join(childComponents...)))
+				return
+			}
+			actionContext.Done(nil)
+		})
 	}
+
 	for _, directory := range directory.Directories {
-		childComponents := append(components, directory.Name)
+		childComponents := append([]string(nil), components...) // We need a copy of our slice to pass to go routine.
+		childComponents = append(childComponents, directory.Name)
 		childDigest, err := digest.NewDerivedDigest(directory.Digest)
 		if err != nil {
 			return util.StatusWrapf(err, "Failed to extract digest for input directory %#v", path.Join(childComponents...))
 		}
 		if err := inputDirectory.Mkdir(directory.Name, 0777); err != nil {
 			return util.StatusWrapf(err, "Failed to create input directory %#v", path.Join(childComponents...))
 		}
-		childDirectory, err := inputDirectory.EnterDirectory(directory.Name)
-		if err != nil {
-			return util.StatusWrapf(err, "Failed to enter input directory %#v", path.Join(childComponents...))
-		}
-		err = d.mergeDirectoryContents(ctx, childDigest, childDirectory, childComponents)
-		childDirectory.Close()
-		if err != nil {
-			return err
-		}
+		directoryName := directory.Name // Ensure we don't grab the reference to an object.
+		actionGroup = append(actionGroup, func(actionContext *re_util.ActionContext) {
+			childDirectory, err := inputDirectory.EnterDirectory(directoryName)
+			if err != nil {
+				actionContext.Done(util.StatusWrapf(err, "Failed to enter input directory %#v", path.Join(childComponents...)))
+				return
+			}
+			// We don't want to close the folder's handle until our children are done, however we also don't
+			// want to hold a slot of the SharedActionQueue (or we might deadlock). Since 'mergeDirectoryContents'
+			// is a blocking function, we can notify the ActionContext that we want to release our hold on a slot
+			// and let a new goroutine block on 'mergeDirectoryContents' for us then call the result handler
+			// when we have finished.
+			doneHandler := actionContext.Defer()
```

Comment:
To me it's not entirely clear what's the use of `actionContext.Defer()`. Couldn't you just call `actionContext.Done()` inside the goroutine below?

Reply:
aaagh, I'm not used to writing go, and I'm still thinking C++ lifetimes apply.